### ISSUE TITLE
Include a claim in both Id token and access token

### ISSUE THREAD

**Issue by arunmurthy** [IdentityServer/IdentityServer3.Samples#241]
**My requirement:** Ability to include a claim in both Id token and access token when requested in a single request

I tried all combinations of ScopeType and alwaysInclude flags and I still can't get it to work.

Looking at https://github.com/IdentityServer/IdentityServer3.Samples/issues/5, I see the following comment by Brock Allen - "there's a Claims property on the Scope class. When the Type is ScopeType.Resource then the claim is always included in the **token**. When the Type is ScopeType.Identity then the claim is available at the user info endpoint, and can optionally be included in the **token** via the alwaysInclude propery on the ScopeClaim."

But when I look at ScopeClaim, I see the following comment for alwaysInclude - "Specifies whether this claim should always be present in the **identity token** (even if an access token has been requested as well). Applies to identity scopes only. Defaults to false."

Looks like Brock meant Id_token when he said "token" in the issue above.

Is there anyway I can get the same claim in both tokens in a single request/response?

**Comment by brockallen** [comment]
You need to define the claim in both the identity scope and the resource scope that the client is requesting.

**Comment by arunmurthy** [comment]
Thanks @brockallen. I defined the claim in both scopes. It doesn't work for me. I must be missing something obvious. My custom scope declaration below...

new Scope
{
    Name = "xyzinfo",
    Description = "xyz information",
    Type = ScopeType.Identity,
    Claims = new List<ScopeClaim>
    {
        new ScopeClaim("xyzName", true)
    }
},
new Scope
{
    Name = "xyzinfo",
    Description = "xyz information",
    Type = ScopeType.Resource,
    Claims = new List<ScopeClaim>
    {
        new ScopeClaim("xyzName")
    }
}

And the results:
1. When I request both id_token and access token, I get "xyzName" claim only in id_token and not in access token
2. When I request only id_token I get "xyzName" claim in the id_token
3. When I request only token I get an "invalid_scope" exception from OpenIdConnect protocol

Basically at no point I am able to get the claim in access token.

**Comment by brockallen** [comment]
I'd check the GetProfileData implementation in your user service -- it should get called once for each token (and there's a context param that indicates which token is being populated).

**Comment by arunmurthy** [comment]
Thanks. It is getting called only once for id_token. I must be missing something in the user service implementation. I will check. Thanks for the quick response. Appreciate it.

**Comment by arunmurthy** [comment]
I still couldn't get this to work. Want to make sure again that I can declare a scope in both identity and resource scopes. For example, scope **xyzinfo** with claim **xyzName** was declared as both resource and identity scope (with alwaysInclude set to true for identity scope). When I do this, depending on the order I declare the scope I get an invalid_scope error when I request either access token or id_token.

For example, if I declare xyzinfo as identity scope first and then as resource scope and in my request if my response type is just "token", I get an invalid_scope. And the same is true vice versa...if I declare xyzinfo as resource scope first and then as identity scope and in my request if my response type is "openid profile xyzinfo", I get an invalid_scope.

**Comment by leastprivilege** [comment]
They need separate names as they model different things - identity vs apis.

**Comment by arunmurthy** [comment]
Got it. Thanks! Let me know if you want me to close this "issue".
